sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,187 @@
+// Package queue  provides Vault plugins with a Priority Queue. It can be used
+// as an in-memory list of queue.Item sorted by their priority, and offers
+// methods to find or remove items by their key. Internally it uses
+// container/heap; see Example Priority Queue:
+// https://golang.org/pkg/container/heap/#example__priorityQueue
+package queue
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+
+	"github.com/mitchellh/copystructure"
+)
+
+// ErrEmpty is returned for queues with no items
+var ErrEmpty = errors.New("queue is empty")
+
+// ErrDuplicateItem is returned when the queue attmepts to push an item to a key that
+// already exists. The queue does not attempt to update, instead returns this
+// error. If an Item needs to be updated or replaced, pop the item first.
+var ErrDuplicateItem = errors.New("duplicate item")
+
+// New initializes the internal data structures and returns a new
+// PriorityQueue
+func New() *PriorityQueue {
+	pq := PriorityQueue{
+		data:    make(queue, 0),
+		dataMap: make(map[string]*Item),
+	}
+	heap.Init(&pq.data)
+	return &pq
+}
+
+// PriorityQueue facilitates queue of Items, providing Push, Pop, and
+// PopByKey convenience methods. The ordering (priority) is an int64 value
+// with the smallest value is the highest priority. PriorityQueue maintains both
+// an internal slice for the queue as well as a map of the same items with their
+// keys as the index. This enables users to find specific items by key. The map
+// must be kept in sync with the data slice.
+// See https://golang.org/pkg/container/heap/#example__priorityQueue
+type PriorityQueue struct {
+	// data is the internal structure that holds the queue, and is operated on by
+	// heap functions
+	data queue
+
+	// dataMap represents all the items in the queue, with unique indexes, used
+	// for finding specific items. dataMap is kept in sync with the data slice
+	dataMap map[string]*Item
+
+	// mapMutex is used to facilitate read/write locks on the dataMap
+	sync.Mutex
```

Comment:
Do we want the lock to be public?  I feel like you should just make this internal and a RWMutex for the Len() call below.

Reply:
I did not, no. Originally it wasn't, the documentation above it is out of date and refers to it's old name `mapMutex`. I've given it a new name `lock` to make it private, and made it a `RWMutex` in c2e9dfdd0